### PR TITLE
Updates for 12. Data Tables

### LINKED ISSUE #284: 12.2 Layout tables test
[12.2 Layout Tables](https://section508coordinators.github.io/ICTTestingBaseline/12DataTables.html#test-instructions-1) should include the "UNLESS..." conditions from b:
c.	Do NOT identify column or row headers using role="columnheader" or role="rowheader" in an ARIA grid (where the data grid is identified using role="grid") 

> UNLESS at least one of the following is true:
> a.	the table has role="presentation"
> b.	the table has role="none"

### LINKED ISSUE #340: Layout tables test instruction 12.-2-1.c: remove role=grid part?
Current instruction:

> Do NOT identify column or row headers using role="columnheader" or role="rowheader" in an ARIA grid (where the data grid is identified using role="grid").

Is column or row headers using role="columnheader" or role="rowheader" problematic even without `role="grid"`? Should 1.c be

'Do NOT identify column or row headers using role="columnheader" or role="rowheader"'?

Related to #284

### LINKED ISSUE #345: 12.1 Update table instructions to indicate explicit role precedence
Table data cell: Check that each data cell uses only one of the following methods 

`<table role="table">` must use aria cell, row, header roles in steps 3 and 4.

### PR DESCRIPTION
**Pull request by kengdoj**
Closes #345

### CODE REVIEW

File: _baselines/12DataTables.md [hunk 1]
```diff
@@ -24,38 +25,49 @@ When `<table>` elements are used for layout purposes, data table structure ele
...
-<p id="12aIC">All content/data visually presented in a table with column and/or row headers where the content is not in a meaningful sequence when linearized.</p>
+<p id="12aIC">Content/data that is visually presented in a table, arranged in rows and columns, where the content is not in a meaningful sequence when linearized.</p>
 
 <p>Note: Linearization of table content is the presentation of a table’s two-dimensional content in one-dimensional order of the content in the source, beginning with the first cell in the first row and ending with the last cell in the last row, from left to right, top to bottom.</p>
 
 #### Test Instructions
 <ol id="12aTI">
-    <li id="12aTI-1">Table: Check that each data table has programmatic markup to identify it as a table using one of the following techniques [SC 1.3.1]:
+    <li>In the following instructions, where an ARIA role is identified, it is the first valid ARIA role attribute value.</li>
+    <li id="12aTI-1">Table: Check the data table has programmatic role of table using one of the following techniques. If more than one technique is used, select one and perform the remaining tests using that selection. [SC 1.3.1]:
```
Comment (kengdoj) [inline _baselines/12DataTables.md:35]:
```suggestion
    <li id="12aTI-1">Table: Check the data table has programmatic role of table using one of the following techniques. If more than one technique is used, select one and perform the remaining tests using that selection. [SC 4.1.2]:
```

**Review by KMSOC — APPROVED** [review]
Looks good. Will make alignment changes to test cases as well

File: _baselines/12DataTables.md [hunk 2]
```diff
@@ -19,76 +20,107 @@ When `<table>` elements are used for layout purposes, data table structure ele
...
 #### Identify Content
-<p id="12aIC">All content/data visually presented in a table with column and/or row headers where the content is not in a meaningful sequence when linearized.</p>
+<p id="12aIC">Content/data that is visually presented in a table, arranged in rows and columns, where the content is not in a meaningful sequence when linearized.</p>
 
 <p>Note: Linearization of table content is the presentation of a table’s two-dimensional content in one-dimensional order of the content in the source, beginning with the first cell in the first row and ending with the last cell in the last row, from left to right, top to bottom.</p>
 
 #### Test Instructions
 <ol id="12aTI">
-    <li id="12aTI-1">Table: Check that each data table has programmatic markup to identify it as a table using one of the following techniques [SC 1.3.1]:
+    <li id="12aTI-1">Table: Check the data table has programmatic role of table using one of the following techniques. If more than one technique is used, select one and perform the remaining tests using that selection. [SC 4.1.2]:
```
Comment (drewnielson) [inline _baselines/12DataTables.md:35]:
```suggestion
    <li id="12aTI-1">Table: Check that the data table has a programmatic role of "table" using one of the following techniques. If more than one technique is used, select the first role explicitly defined and perform the remaining tests using that selection. [SC 4.1.2]:
```

File: _baselines/12DataTables.md [hunk 3]
```diff
@@ -19,76 +20,107 @@ When `<table>` elements are used for layout purposes, data table structure ele
...
         <ul>
-            <li>HTML <code>&lt;table&gt;</code></li>
+            <li>HTML <code>&lt;table&gt;</code> that does not have a role attribute that changes its role from table, such as <code>role="presentation"</code> or <code>role="none"</code>.</li>
```
Comment (drewnielson) [inline _baselines/12DataTables.md:37]:
```suggestion
            <li>HTML <code>&lt;table&gt;</code> that does not have an explicitly defined role attribute that changes its role from table, such as <code>role="presentation"</code> or <code>role="none"</code>.</li>
```

File: _baselines/12DataTables.md [hunk 4]
```diff
@@ -19,76 +20,107 @@ When `<table>` elements are used for layout purposes, data table structure ele
...
             <li>ARIA <code>role="table"</code></li>
             <li>ARIA <code>role="grid"</code></li>
-        </ul></li>
-    <li id="12aTI-2">Check that the data <code>&lt;table&gt;</code> element does <strong>NOT</strong> have <code>role="presentation"</code> or <code>role="none"</code>. [SC 1.3.1]</li>
-    <li id="12aTI-3">Table data cell: Check that each data cell uses only one of the following methods to identify it as a data cell within a table row depending on the technique identified in the first step [SC 1.3.1]:
+            <li>ARIA <code>role="treegrid"</code></li>
+        </ul>
+    </li>
+    <li id="12aTI-2">Table data cell: For the technique used, check that each data cell is programmatically assigned a role of cell and within a row [SC 4.1.2]:
```
Comment (drewnielson) [inline _baselines/12DataTables.md:43]:
```suggestion
    <li id="12aTI-2">Table data cell: For the technique used, check that each data cell is programmatically assigned a role of cell and within a parent element with a role of row [SC 4.1.2]:
```

File: _baselines/12DataTables.md [hunk 5]
```diff
@@ -19,76 +20,107 @@ When `<table>` elements are used for layout purposes, data table structure ele
...
-            </ul></li>
-        <li>For any HTML <code>&lt;table&gt;</code> that uses <strong>BOTH</strong> <code>scope</code> <strong>AND</strong> refers to header IDs using <code>&lt;td headers&gt;</code> attributes in the same table, any data cell with a <code>headers</code> reference will override any <code>scope</code> attributes for associated table headers for that particular data cell. Therefore, data cells with a <code>headers</code> reference, must identify all relevant headers, independent from and regardless of <code>scope</code> attributes in associated headers.</li>
-        <li>For ARIA <code>role="table"</code>: each column header must have <code>role="columnheader"</code> and each row header must have <code>role="rowheader"</code>.</li>
-        <li>For ARIA <code>role="grid"</code>: each column header must have <code>role="columnheader"</code> and each row header must have <code>role="rowheader"</code> (if the ARIA grid is not making use of the native HTML <code>&lt;table&gt;</code> element and structure).</li>
-	</ul></li>
+            <li>For HTML <code>&lt;table&gt;</code>: the <code>&lt;td&gt;</code> cell must be within a <code>&lt;tr&gt;</code> row.</li>
```
Comment (drewnielson) [inline _baselines/12DataTables.md:45]:
```suggestion
            <li>For HTML <code>&lt;table&gt;</code>: the <code>&lt;td&gt;</code> cell must be within a parent <code>&lt;tr&gt;</code> row element.</li>
```

File: _baselines/12DataTables.md [hunk 6]
```diff
@@ -19,76 +20,107 @@ When `<table>` elements are used for layout purposes, data table structure ele
...
+            <li>For ARIA <code>role="table"</code>: the data cell element with <code>role="cell"</code> must be within an element with <code>role="row"</code>.</li>
+            <li>For ARIA <code>role="grid"</code> or ARIA <code>role="treegrid"</code>: the data cell element with <code>role="gridcell"</code> is within an element with <code>role="row"</code>.</li>
```
Comment (drewnielson) [inline _baselines/12DataTables.md:47]:
```suggestion
            <li>For ARIA <code>role="grid"</code> or ARIA <code>role="treegrid"</code>: the data cell element with <code>role="gridcell"</code> is within a parent element with <code>role="row"</code>.</li>
```

File: _baselines/12DataTables.md [hunk 7]
```diff
@@ -19,76 +20,107 @@ When `<table>` elements are used for layout purposes, data table structure ele
...
+        </ul>
+    </li>
+    <li id="12aTI-3">Table header cell: For the technique used, check that each header cell is programmatically assigned a role of header and within a row [SC 4.1.2]:
```
Comment (drewnielson) [inline _baselines/12DataTables.md:50]:
```suggestion
    <li id="12aTI-3">Table header cell: For the technique used, check that each header cell is programmatically assigned a role of header and within a parent row element [SC 4.1.2]:
```

File: _baselines/12DataTables.md [hunk 8]
```diff
@@ -19,76 +20,107 @@ When `<table>` elements are used for layout purposes, data table structure ele
...
-        <li>For ARIA <code>role="grid"</code>: each column header must have <code>role="columnheader"</code> and each row header must have <code>role="rowheader"</code> (if the ARIA grid is not making use of the native HTML <code>&lt;table&gt;</code> element and structure).</li>
-	</ul></li>
+            <li>For HTML <code>&lt;table&gt;</code>: the <code>&lt;td&gt;</code> cell must be within a parent <code>&lt;tr&gt;</code> row element.</li>
+            <li>For ARIA <code>role="table"</code>: the data cell element with <code>role="cell"</code> must be within a parent element with <code>role="row"</code>.</li>
+            <li>For ARIA <code>role="grid"</code> or ARIA <code>role="treegrid"</code>: the data cell element with <code>role="gridcell"</code> is within a parent element with <code>role="row"</code>.</li>
+        </ul>
+    </li>
+    <li id="12aTI-3">Table header cell: For the technique used, check that each header cell is programmatically assigned a role of header and within a parent row element [SC 4.1.2]:
+        <ul>
+            <li>For HTML <code>&lt;table&gt;</code>: the <code>&lt;th&gt;</code> header must be within a <code>&lt;tr&gt;</code> row.</li>
```
Comment (drewnielson) [inline _baselines/12DataTables.md:52]:
```suggestion
            <li>For HTML <code>&lt;table&gt;</code>: the <code>&lt;th&gt;</code> header must be within a parent <code>&lt;tr&gt;</code> row element.</li>
```

File: _baselines/12DataTables.md [hunk 9]
```diff
@@ -19,76 +20,107 @@ When `<table>` elements are used for layout purposes, data table structure ele
...
+            <li>For an ARIA <code>role="table"</code>, ARIA <code>role="grid"</code> or <code>role="gridtree"</code>: each element with <code>role="columnheader"</code> or <code>role="rowheader"</code> must be within an element with <code>role="row"</code>.</li>
```
Comment (drewnielson) [inline _baselines/12DataTables.md:53]:
```suggestion
            <li>For an ARIA <code>role="table"</code>, ARIA <code>role="grid"</code> or <code>role="gridtree"</code>: each element with <code>role="columnheader"</code> or <code>role="rowheader"</code> must be within a parent element with <code>role="row"</code>.</li>
```

File: _baselines/12DataTables.md [hunk 10]
```diff
@@ -19,76 +20,107 @@ When `<table>` elements are used for layout purposes, data table structure ele
...
+**Baseline Test ID:** 12.B-DataTableHeaderAssociation
+
+#### Identify Content
+<p id="12bIC">For any data table identified in 12.A, identify all column and row headers for each data cell.</p>
+
+<ol>
+    <li id="12bTI-1">Data cell to header(s) association: Use the programmatic technique (HTML or ARIA) identified in 12.A. Check that data cells are programmatically associated with its headers [SC 1.3.1]:
+        <ul>
+            <li id="12bTI-1a">For a simple HTML <code>&lt;table&gt;</code> only: with column headers only in the first row that apply to all data cells in the same column and row headers only in the first column that apply to all data cells in the same row, each header cell may be marked up with <code>&lt;th&gt;</code> without additional attributes.</li>
+            <li id="12bTI-1b">For any HTML <code>&lt;table&gt;</code>: a header cell may be marked up with <code>&lt;th scope="col|row|colgroup|rowgroup"&gt;</code> where the data cells that the headers apply to are in the same column, row, column group, or row group, respectively.
```
Comment (drewnielson) [inline _baselines/12DataTables.md:71]:
```suggestion
            <li id="12bTI-1b">For any HTML <code>&lt;table&gt;</code>: a header cell may be marked up with <code>&lt;th scope="[col|row|colgroup|rowgroup]"&gt;</code> where the data cells that the headers apply to are in the same column, row, column group, or row group, respectively.  
```

File: _baselines/12DataTables.md [hunk 11]
```diff
@@ -19,76 +20,107 @@ When `<table>` elements are used for layout purposes, data table structure ele
...
+#### Identify Content
+<p id="12bIC">For any data table identified in 12.A, identify all column and row headers for each data cell.</p>
+
+<ol>
+    <li id="12bTI-1">Data cell to header(s) association: Use the programmatic technique (HTML or ARIA) identified in 12.A. Check that data cells are programmatically associated with its headers [SC 1.3.1]:
+        <ul>
+            <li id="12bTI-1a">For a simple HTML <code>&lt;table&gt;</code> only: with column headers only in the first row that apply to all data cells in the same column and row headers only in the first column that apply to all data cells in the same row, each header cell may be marked up with <code>&lt;th&gt;</code> without additional attributes.</li>
+            <li id="12bTI-1b">For any HTML <code>&lt;table&gt;</code>: a header cell may be marked up with <code>&lt;th scope="col|row|colgroup|rowgroup"&gt;</code> where the data cells that the headers apply to are in the same column, row, column group, or row group, respectively.  
+                <ul>
+                    <li>In HTML4, <code>&lt;td scope&gt;</code> is supported.</li>
```
Comment (drewnielson) [inline _baselines/12DataTables.md:73]:
```suggestion
                    <li>In HTML4, <code>&lt;td scope="[row|col]"&gt;</code> is supported.</li>
```

File: _baselines/12DataTables.md [hunk 12]
```diff
@@ -19,76 +20,107 @@ When `<table>` elements are used for layout purposes, data table structure ele
...
+                    <li>In HTML5, <code>&lt;td scope&gt;</code> is not supported so all header cells must be <code>&lt;th&gt;</code>. </li>
```
Comment (drewnielson) [inline _baselines/12DataTables.md:74]:
```suggestion
                    <li>In HTML5, <code>&lt;td scope="[row|col]"&gt;</code> is not supported so all header cells must be <code>&lt;th&gt;</code>. </li>
```

File: _baselines/12DataTables.md [hunk 13]
```diff
@@ -19,76 +20,107 @@ When `<table>` elements are used for layout purposes, data table structure ele
...
+            <li id="12bTI-1b">For any HTML <code>&lt;table&gt;</code>: a header cell may be marked up with <code>&lt;th scope="[col|row|colgroup|rowgroup]"&gt;</code> where the data cells that the headers apply to are in the same column, row, column group, or row group, respectively.  
+                <ul>
+                    <li>In HTML4, <code>&lt;td scope="[row|col]"&gt;</code> is supported.</li>
+                    <li>In HTML5, <code>&lt;td scope="[row|col]"&gt;</code> is not supported so all header cells must be <code>&lt;th&gt;</code>. </li>
+                    <li>The <code>scope</code> only applies to cells that occur after the header cell(s) in the reading order.</li>
+                </ul>
+            </li>
+            <li id="12bTI-1c">For any HTML <code>&lt;table&gt;</code>: each <code>&lt;td&gt;</code> data cell may be associated to header cell(s) with <code>&lt;td headers&gt;</code>:
```
Comment (drewnielson) [inline _baselines/12DataTables.md:78]:
```suggestion
            <li id="12bTI-1c">For any HTML <code>&lt;table&gt;</code>: each <code>&lt;td&gt;</code> data cell may be associated to header cell(s) with <code>&lt;td headers="[id]"&gt;</code>:
```

File: _baselines/12DataTables.md [hunk 14]
```diff
@@ -19,76 +20,107 @@ When `<table>` elements are used for layout purposes, data table structure ele
...
+                <ul>
```
Comment (drewnielson) [inline _baselines/12DataTables.md:79]: Wouldn't the headers technique also apply to ARIA tables? Should we repeat these bullets for the ARIA approach or modify the parent bullet to apply to both HTML tables and ARIA tables?

Comment (kengdoj) [reply _baselines/12DataTables.md:79]: I don't think so; https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/table_role

`role=rowheader` or `role=columnheader` are to be used in an aria table, grid, or treegrid.

**Review by drewnielson — APPROVED** [review]
Looks great

File: _baselines/12DataTables.md (hunk 1 above)
Comment (cliffbert) [reply _baselines/12DataTables.md]: Are we using "Check that" or "Check"?

**Pull request merged**